sssom/util.py

```diff
@@ -1505,3 +1512,84 @@ def _get_sssom_schema_object() -> SSSOMSchemaView:
         else SSSOMSchemaView()
     )
     return sssom_sv_object
+
+
+def flip_mappings(
+    df: pd.DataFrame, subject_prefix: str, merge_flipped: bool = True
+) -> pd.DataFrame:
+    """Switching subject and objects based on their prefixes and adjusting predicates accordingly.
+
+    :param df: Pandas dataframe.
+    :param subject_prefix: Prefix of subjects desired.
+    :param merge_flipped: If True (default), add flipped dataframe to input else,
+                          just return flipped data.
+    :return: Pandas dataframe with all subject IDs having the same prefix.
+    """
+    subject_starts_with_prefix_condition = df[SUBJECT_ID].str.startswith(
+        subject_prefix + ":"
+    )
+    object_starts_with_prefix_condition = df[OBJECT_ID].str.startswith(
+        subject_prefix + ":"
+    )
+    blank_predicate_modifier = df[PREDICATE_MODIFIER] == ""
+    predicate_flip_map = PREDICATE_FLIP_DICTIONARY
+
+    # predicate_modified_df = pd.DataFrame(df[not_predicate_modifier])
+    non_predicate_modified_df = pd.DataFrame(df[blank_predicate_modifier])
+
+    prefixed_subjects_df = pd.DataFrame(
+        non_predicate_modified_df[
+            (
+                subject_starts_with_prefix_condition
+                & ~object_starts_with_prefix_condition
+            )
+        ]
+    )
+    non_prefix_subjects_df = pd.DataFrame(
+        non_predicate_modified_df[
+            (
+                ~subject_starts_with_prefix_condition
+                & object_starts_with_prefix_condition
+            )
+        ]
+    )
+    df_to_flip = non_prefix_subjects_df.loc[
+        non_prefix_subjects_df[PREDICATE_ID].isin(list(predicate_flip_map.keys()))
+    ]
+    list_of_subject_object_columns = [
+        x for x in df_to_flip.columns if x.startswith(("subject", "object"))
+    ]
+    flipped_df = df_to_flip.rename(
+        columns=_flip_column_names(list_of_subject_object_columns)
+    )
+    flipped_df = flipped_df[df.columns]
+    flipped_df[PREDICATE_ID] = flipped_df[PREDICATE_ID].apply(
```

Comment:
I think we also need to attach additional metadata:

- [ ] set https://mapping-commons.github.io/sssom/mapping_source/ (`flipped_df[MAPPING_SOURCE]`) to the `mapping_set_id` of the input df.
- [ ] set `flipped_df[MAPPING_JUSTIFICATION]` to "sempav:FlippedMapping" (see https://github.com/mapping-commons/semantic-mapping-vocabulary/pull/23 for context)

Reply:
```suggestion
    flipped_df[PREDICATE_ID] = flipped_df[PREDICATE_ID].map(predicate_flip_map)
```

Reply:
Applying this change @cthoyt will break the code, but yes the syntax seems more terse. Thanks, I didn't know you can do it this way!